SI-6375, warn on lost annotation.

paulp opened the pull request:
Annotations on abstract vals which are not meta-annotated
were silently discarded. Still discarded, only less silently.

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1215/

scala-jenkins commented:
jenkins job pr-rangepos: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1215/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1931/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1931/

paulp commented:
PLS REBUILD ALL

paulp commented:
Failed to retarget the checkfile to master.

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1218/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1218/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/1933/

xeno-by commented on test/files/neg/t6375.check:
```diff
@@ -0,0 +1,9 @@
+t6375.scala:6: warning: Annotation is unused - you may need to add a meta-annotation (see scala.annotation.meta)
```
How about making the warning more specific, e.g. `Annotation is discarded - use the @getter meta-annotation to preserve it`? 

Also "scala meta annotation" seems excellently googlable. Maybe we could omit the reference to `scala.annotation.meta`?

JamesIry commented:
ping @paulp

paulp committed:
Cleaned up meta-annotations.

There were some missing and some others not added to the
set of meta-annotations. I added the missing ones and
defined the set in such a way as to include them all, even
those which have not yet been born.

paulp committed:
SI-6375, warn on lost annotation.

Annotations on abstract vals which are not meta-annotated
were silently discarded. Still discarded, only less silently.
I warned on as many "lost annotation" situations as I was
reasonably able to cover without false positives.

paulp commented:
That sucked me in for about half a day longer than I'd intended.

paulp committed:
Tweaked meta-annotation error based on feedback.

xeno-by commented on src/compiler/scala/tools/nsc/typechecker/MethodSynthesis.scala:
```diff
@@ -171,14 +171,46 @@ trait MethodSynthesis {
+      val targetClass   = defaultAnnotationTarget(tree)
+      val retained      = deriveAnnotations(annotations, targetClass, keepClean = true)
+
+      annotations filterNot (retained contains _) foreach (ann => issueAnnotationWarning(tree, ann, targetClass))
```
What do you think about `annotations diff retained`?

xeno-by replied:
Also going through retained seems a bit roundabout, because retained is essentially a result of `annotations filter <some predicate from deriveAnnotations>`. If only we had a method like deriveAnnotations, but with the predicate inverted.

xeno-by commented on src/compiler/scala/tools/nsc/typechecker/MethodSynthesis.scala:
```diff
@@ -171,14 +171,46 @@ trait MethodSynthesis {
+        // may need to make their way to parameters of the constructor as
+        // well as fields of the class, etc.
+        if (!mods.isParamAccessor) annotations foreach (ann =>
+          if (!trees.exists(_.symbol hasAnnotation ann.symbol))
```
Would be great if we had a more robust way of determining that we're looking at `ann`, but I assume that's one of the pain points you mentioned in the commit message. In any case, this patch is a definite improvement over the state of the art, so lgtm.

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1383/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1383/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2095/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Failed - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2095/

paulp commented:
Looks like the usual bogus baloney.

paulp commented:
PLS REBUILD ALL

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1389/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1389/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2103/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2103/
